### ISSUE TITLE
【Mosaic bug】   incorrect image segmentation affects training

### ISSUE THREAD

**Issue by sizhongyibanhts** [ultralytics/ultralytics#21925]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Problem Phenomenon（问题现象）
Mosaic is a highly effective training method, but I've noticed some problems with it.（mosaic是非常有效的训练手段，但我发现了一些问题）。You can visit this page to see a more complete analysis of the [problem.]( https://blog.csdn.net/weixin_37763484/article/details/151152493)

I encountered an issue with mosaic augmentation, where it incorrectly splits data and generates illogical bounding boxes. Today, while training an insulator recognition model, I found this problem to be severe—almost every ten images contains a serious error. In these erroneous cases, the bounding boxes do not contain any insulators at all, but instead only include irrelevant background regions or other electrical equipment.（我遇到了一个关于mosaic增强的问题，mosaic会错误的切分数据，产生不符合逻辑的标注框。今天我在训练绝缘子识别模型，发现这个现象很严重，几乎每十张图片就会出现一个严重的错误。这种错误的标注框中，根本不包含绝缘子，只包含一些无关的背景或其他电力设备。）

The original image is as follows（原始图像如下）：

<img width="823" height="969" alt="Image" src="https://github.com/user-attachments/assets/2fed5dc2-df6b-4bb5-ae9c-a8ffeff4caa1" />

After applying mosaic data augmentation（使用mosaic数据增强后）：

<img width="1130" height="660" alt="Image" src="https://github.com/user-attachments/assets/479494b1-2125-44fd-aefb-e724bce4e3c3" />

### Problem Analysis 问题分析

If you look closely at the insulator in the bottom right corner of the augmented image, the position of the bounding box is clearly offset（你可以仔细观察，增强之后图像右下角的绝缘子，标注框位置明显偏移）。

This would likely have a severe impact on the model's training. There are even more severe cases where the insulator is no longer visible within the augmented bounding box at all. This happens because my target object is located along the diagonal of its bounding box. When the mosaic augmentation process truncates the box, it's highly likely to crop a diagonal region that doesn't actually contain the target（这应该会严重影响到模型训练，还有一些更严重的情况，增强后的标注框中已经无法看到绝缘子。这是因为我的目标位于标注框的对角线上，当你的mosaic增强把坐标框截断的时候，非常有可能截取到不包含目标的对角线区域。）

### 解决建议

When I turned off mosaic, similar issues became much rarer (almost non-existent), so mosaic must be too aggressive.（（当关闭马赛克后，类似的情况就少了很多（几乎没有），因此应该是马赛克过于粗暴了。）

Since my targets are diagonal bounding boxes, mosaic cropping needs to recalculate the new coordinates based on the intersection of the diagonal and the cropping boundary.（由于我的目标是斜跨标注框对角线的，因此在马赛克截取的时候，需要重新根据对角线位置计算新的坐标框（计算截取位置与对角线的交点）。）

Alternatively, I should either not use mosaic for my scenario, or ensure that mosaic augmentation completely preserves the bounding boxes for certain special classes.（或者在我的场景下，不要使用mosaic增强，或者保证马赛克增强的时候，把某些特殊类别的坐标框完整的保留。）

Another option is to use mosaic less or turn it off early. However, the model's confusion and wasted training that happens while it's on would be hard to correct.（或者少用一些马赛克，提前关闭马赛克，但是未关闭的时候，造成的模型误会和训练浪费，恐怕难以纠正。）

Or, when mosaic crops more than 50% of a target's area, automatically black out the remaining part.（或者当马赛克将某个目标截取了超过50%面积的时候，自动将目标剩余的部分涂黑。）

### Environment

windows11  

This bug is not related to the computer system.

这个bug与计算机系统无关

### Minimal Reproducible Example

你可以在这个网页中，看到问题的截图[my blog](https://blog.csdn.net/weixin_37763484/article/details/151152493)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sizhongyibanhts, thank you for your interest in Ultralytics 🚀! This is an automated response to help get things moving quickly. An Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. For mosaic-related issues, please include:
-Exact ultralytics version: run `python -c "import ultralytics, torch; print('ultralytics', ultralytics.__version__, 'torch', torch.__version__)"`  
-Task/model you used (e.g., detection with YOLO11n)  
-Full training command or Python snippet reproducing the issue  
-A small sample (2–4) of images + labels that demonstrate the problematic mosaic result  
-The exact data.yaml and any custom hyp/args affecting augmentation (mosaic probability, scale/translate/degrees/shear/perspective, mixup/cutout/copy_paste, imgsz, rect, batch, workers, seed)  
-Screenshots of saved train_batch*.jpg (mosaic previews) from runs where the issue occurs  
-Confirmation whether the problem persists with a fixed seed (e.g., `--seed 0`) and without mixed augmentations (e.g., disabling mixup/copy_paste)  

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To speed up triage, please share the MRE elements above along with a short code snippet that reproduces the mosaic behavior, for example:

```python
from ultralytics import YOLO

model = YOLO('yolo11n.pt')  # or your exact model
model.train(
    data='data.yaml',
    epochs=1,
    imgsz=640,
    batch=8,
    workers=0,
    seed=0,
    mosaic=1.0,  # specify the value you use
    mixup=0.0,
    copy_paste=0.0,
    project='repro',
    name='mosaic_issue',
)
```

💡 Tip: please attach the corresponding train_batch*.jpg from repro/mosaic_issue to illustrate the misaligned boxes.

**Comment by glenn-jocher** [comment]
Thanks for the thorough analysis—this is a known side-effect of Mosaic with axis-aligned boxes: after the affine step we clip boxes and filter by area ratio (see the RandomPerspective box filter, `area_thr=0.1`) which can leave little or no foreground when objects run along a diagonal; details in the [Mosaic augmentation guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/#mosaic-mosaic) and the [transform/box-candidate reference](https://docs.ultralytics.com/reference/data/augment/#box_candidates-staticmethod). To mitigate, reduce Mosaic probability and disable it early, or switch to OBB so boxes rotate with objects (see the [OBB task docs](https://docs.ultralytics.com/tasks/obb/)); for example:
```bash
yolo train model=yolo11n.pt data=insulator.yaml epochs=100 mosaic=0.2 close_mosaic=50 mixup=0.1 cutmix=0.1
```
If you still suspect a bug, please attach a tiny reproducible sample (4–8 images with labels) plus your exact train command and `yolo checks` output (text, not screenshots), and we’ll investigate or you can open a new issue for visibility.

**Comment by sizhongyibanhts** [comment]
Will using an OBB model reduce model accuracy? Alternatively, can I modify the Mosaic function myself to ensure that Mosaic does not excessively cut the target bounding boxes?

**Comment by glenn-jocher** [comment]
Good questions. OBB doesn’t inherently reduce accuracy; for diagonally oriented targets it often improves localization when you train and evaluate with OBB labels and metrics—see this brief [OBB overview](https://www.ultralytics.com/blog/what-is-oriented-bounding-box-obb-detection-a-quick-guide). For a quick test:
```bash
yolo obb train model=yolo11n-obb.pt data=insulator-obb.yaml epochs=100 imgsz=640
```
If you prefer AABB, a minimal safeguard is to drop heavily truncated boxes after the affine step by raising the area filter in `ultralytics/data/augment.py` inside `RandomPerspective.__call__`:
```python
# replace
i = self.box_candidates(box1=instances.bboxes.T, box2=new_instances.bboxes.T, area_thr=0.10)
# with, e.g.
i = self.box_candidates(box1=instances.bboxes.T, box2=new_instances.bboxes.T, area_thr=0.30)
```
You can also reduce geometric motion to limit cuts:
```bash
yolo train model=yolo11n.pt data=insulator.yaml epochs=100 translate=0.02
```
If you need format conversions for OBB↔AABB during evaluation or tooling, the box utilities are referenced here: [OBB box conversion ops](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.xywhr2xyxyxyxy).
